### ISSUE TITLE
Use Widget in the Create edit form

### ISSUE THREAD

**Issue by clarkjefcoat** [ozone-development/ozp-center#287]
Can a setting be added that sets an app to launch in a new tab by default?  Currently there is a per-user setting in HUD that can specify where all apps launch.  Can this setting be made on a per-app basis?  It was mentioned that this was a capability in the legacy software.

If we decide to do this it will require additional tickets in HUD, center and webtop.

**Comment by mparizernc** [comment]
In legacy it was handled through a flag in the listing.  (Open in New Tab) If the flag was true the listing would open in a new tab if it was false it would at it to OWF and open it there.

**Comment by sdeighan** [comment]
I have some questions about this:
Do you mean a setting that is defined by a user of the application or the listing itself?
Also, if a user sets a global user setting will it override a listing default?

There are some concerns with a by application default setting and a general default setting. If we're concerned with just matching legacy there will be some clashing. Since we have a settings panel and there are no other application level settings, we will need to integrate the user-defined app-specific defaults into the global settings panel or there will be significant confusion. This will take up a lot of space but it's the only place where it will be clear. It will be with the global default setting so that, when someone goes to change any of them, both types of settings will be obvious.

We will also need to have three states, "Default" so that, by default, the application would open with whatever default. The other two options would need to be the two methods themselves, the Data analysis framework and in a new tab. We couldn't just say "Not default" or similar because if the user changes the default it would flip all of their settings.

Now, this brings us to the listing-chosen defaults. If we implement those, decisions will need to be made whether these additional listing-specific defaults can be changed or overridden by the user. We have been going forward with features and such with the idea that the users should be allowed to customize. If they are allowed, we might be able to just populate them into the listing-specific preference, but that becomes confusing because the user did not set them. There is the potential for an application to behave against all defaults except one that a user did not set and, in order to fix it, they will have to go into a panel where they didn't add a behavior and fix it.

If they are not allowed to go against the default chosen by a listing, we will be locking all the users of that listing into an opening method that may or may not be compatible with how they work. Additionally, it will be confusing for users because of all the other levels of app-opening default since this would override all the default preferences.

In conclusion, the default opening preference paradigm was shifted with the new system. We certainly can add another level, but it will add another layer of complexity for our users. Adding two will increase the complexity of the defaults roughly by a square rather than a simple multiplier. In order to stay with the same design goals, I would suggest adding users' app-specific preferences but not app-specific settings from the listings unless it is somehow approved by a user when it's added to their bookmarks. Listing-set app-specific preferences will compete with all the other defaults and be difficult from both a policy and a UI standpoint.

**Comment by clarkjefcoat** [comment]
This request is for a setting that is defined by the listing itself.  This would override any global user setting.  The way I understood the request is how you explained in your next to last paragraph ("If they are not allowed...").

An app that is being proposed for production does not currently play well with webtop.  I think that is the reason behind this request.  That issue is still being worked.  Let's wait to see how that turns out before we spend too much time with this.

As you noted this request would complicate the design and implementation of the system.

**Comment by tonyalletag** [comment]
This is something we will have to implement with the new back end. It was rank as the #1 priority for the CSWG by multiple users and content stewards.

When submitting and App the App owner wants to set the default launch method to not launch to the webtop.

@bhalt Would it be better to just set all Apps by default to launch in a new tab?

**Comment by mparizernc** [comment]
@tonyalletag Just as an FYI this will require front end changes to the HUD and the user settings for both Center and HUD, since users have the option to open either way in both places.  Also does this mean if the default is set that the user can override the default setting for the listing?

**Comment by bhalt** [comment]
@tonyalletag Just to clarify, currently applications launched from Center appear in a new tab, with no option otherwise. Applications launched from HUD are the only ones that are by default launched in a new tab, but are affected by the change in the user setting.

Personally, I think it's a bit clunky to "launch" HUD apps into Webtop as is now, because it automatically adds it to your original Dashboard. If it is actually adding to the dashboard, instead of just launching, that should be more clear (maybe language of "Add to Webtop" with an alert that says "Would you like to add this to a new Dashboard or one of your current Dashboards?" with a select field). I'm not sure that setting all Apps by default to launch in a new tab would solve it, because another org is going to someday have a Webtop-only app and want the opposite.

I don't think "launch location" is worth a default setting. Not only is it not consistent throughout the platform, but it will probably be much more app-dependent as we go forward. I would recommend taking away the user setting for now (especially since the entire Settings modal could then be removed) and letting the user select "Open in New Tab" or "Add to Webtop" from the HUD app menus as they can now. For apps that the developer specifies can only be one or the other, the other option would be removed from that app's HUD menu. Default launch should match whatever Center is (currently new tab). Thoughts?

**Comment by bhalt** [comment]
@tonyalletag Just to clarify, currently applications launched from Center appear in a new tab, with no option otherwise. Applications launched from HUD are the only ones that are by default launched in a new tab, but are affected by the change in the user setting.

Personally, I think it's a bit clunky to "launch" HUD apps into Webtop as is now, because it automatically adds it to your original Dashboard. If it is actually adding to the dashboard, instead of just launching, that should be more clear (maybe language of "Add to Webtop" with an alert that says "Would you like to add this to a new Dashboard or one of your current Dashboards?" with a select field). I'm not sure that setting all Apps by default to launch in a new tab would solve it, because another org is going to someday have a Webtop-only app and want the opposite.

I don't think "launch location" is worth a default setting. Not only is it not consistent throughout the platform, but it will be overridden by this required developer feature, and it will probably be much more app-dependent as we go forward. I would recommend taking away the user setting for now (especially since the entire Settings modal could then be removed) and letting the user select "Open in New Tab" or "Add to Webtop" from the HUD app menus (similar to what is there now). For apps that the developer specifies can only be one or the other, the other option would be removed from that app's HUD menu. If the developer allows both, default launch should match what Center is (new tab). If the developer allows only Webtop, the alert should read "This app can only be opened through Webtop. Would you like to add this to a new Dashboard or one of your current Dashboards?". Thoughts?

Side note: Should these "only open in new tab" apps be blocked from adding to Webtop from Center or Webtop? This may need a larger conversation.

**Comment by tonyalletag** [comment]
Lets talk about this at the next iteration planning

**Comment by bhalt** [comment]
Solution discussed and agreed upon -- we are removing the ability to open apps from HUD into Webtop (currently broken), removing the setting to change the default launch (currently in the setting modal), and removing the setting modal on the global menu.

**Comment by tonyalletag** [comment]
updated ticket title

**Comment by bhalt** [comment]
UPDATED TASKING FOR TICKET: 

1. Create a Webtop-enabled setting for each app to opt into
2. Webtop-enabled toggle button should be added to the Center "Submit a Listing" form, on the "Listing Details" section before "Singleton" toggle button with the text:

##Webtop-enabled
Allow the application to be opened into an iframe inside a Webtop dashboard. 
*Toggle ON text:* "The application can be launched into Webtop"
*Toggle OFF text:* "The application can only be launched into a new browser tab"

The "Singleton" option below it should be disabled if user does not choose to Webtop-enable their app.

**Comment by bhalt** [comment]
With the recent decision to merge Widget listing types with the Webtop-enabled tag -- we have new implementation for the Center "Submit a Listing" form. There should no longer be a "Webtop-enabled" toggle, but a note should be shown underneath the Listing Type select.
![image](https://cloud.githubusercontent.com/assets/8047457/13995994/23326362-f102-11e5-9dfe-0f00ae93d877.png)

Code: `<p class="small"><b>Note:</b> A widget is the only listing type that can be displayed on the Webtop.</p>`

Ideally, we would also only show the Singleton toggle if a user selects "Widget"

@SOAP1914

**Comment by tonyalletag** [comment]
This is mostly done but updates need to be implemented on the create edit form.
